Mathlib/CategoryTheory/NatTransCartesian.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2025 Sina Hazratpour. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sina Hazratpour, Wojciech Nawrocki
+-/
+module
+
+public import Mathlib.CategoryTheory.Functor.TwoSquare
+public import Mathlib.CategoryTheory.Limits.Shapes.Pullback.CommSq
+
+/-!
+# Cartesian natural transformations
+
+A natural transformation between functors is cartesian if all its naturality squares are
+pullback squares.
+
+## Main results
+
+- `NatTrans.isCartesian_of_discrete` shows that any natural transformation between functors from a
+  discrete category is cartesian.
+
+- `NatTrans.isCartesian_of_isIso` shows that any natural isomorphism is cartesian.
+
+- `NatTrans.isIso_of_isCartesian_of_isIso_app_terminal` shows that a cartesian natural
+  transformation is an isomorphism if its component at a terminal object is an isomorphism.
+
+- `NatTrans.isCartesian_of_isPullback_isTerminal_from` shows that a natural transformation is
+  cartesian if all its naturality squares to the terminal object are pullback squares.
+
+- `NatTrans.IsCartesian.comp` shows that the composition of two cartesian natural transformations is
+
+### References
+
+This material is adapted from the polynomial functors project developed during the
+Trimester "Prospects of Formal Mathematics" at the Hausdorff Institute (HIM) in Bonn:
+
+https://github.com/sinhp/Poly
+
+-/
+
+@[expose] public section
+
+universe v' u' v u
+
+open CategoryTheory Limits IsPullback
+
+namespace CategoryTheory
+
+variable {J : Type v'} [Category.{u'} J] {C : Type u} [Category.{v} C]
+variable {K : Type*} [Category K] {D : Type*} [Category D]
+
+namespace NatTrans
+
+open Functor
+
+/-- A natural transformation is cartesian if all its naturality squares are pullbacks. -/
+def IsCartesian {F G : J ⥤ C} (α : F ⟶ G) : Prop :=
```

Comment:
This should probably be made a typeclass as they are many lemmas below which could be instances.